housekeeping/src/main/groovy/whelk/housekeeping/NotificationSender.groovy

```diff
@@ -0,0 +1,428 @@
+import org.simplejavamail.email.EmailBuilder
+import org.simplejavamail.mailer.MailerBuilder
+import whelk.Document
+import whelk.JsonLd
+import whelk.Whelk
+import groovy.transform.CompileStatic
+import groovy.util.logging.Log4j2 as Log
+import whelk.util.PropertyLoader
+
+import java.sql.Connection
+import java.sql.PreparedStatement
+import java.sql.ResultSet
+import java.sql.Timestamp
+import java.time.Instant
+import java.time.ZoneOffset
+import java.time.ZonedDateTime
+import java.time.format.DateTimeFormatter
+import java.time.temporal.ChronoUnit
+
+@CompileStatic
+@Log
+class NotificationSender extends HouseKeeper {
+
+    private final String STATE_KEY = "Email notifications"
+    private String status = "OK"
+    private final Whelk whelk
+    private final Mailer mailer
+    private final String senderAddress
+
+    public NotificationSender(Whelk whelk) {
+        this.whelk = whelk
+        Properties props = PropertyLoader.loadProperties("secret")
+        if (props.containsKey("smtpServer") &&
+                props.containsKey("smtpPort") &&
+                props.containsKey("smtpSender") &&
+                props.containsKey("smtpUser") &&
+                props.containsKey("smtpPassword"))
+            mailer = MailerBuilder
+                .withSMTPServer(
+                        (String) props.get("smtpServer"),
+                        Integer.parseInt((String)props.get("smtpPort")),
+                        (String) props.get("smtpUser"),
+                        (String) props.get("smtpPassword")
+                ).buildMailer()
+        senderAddress = props.get("smtpSender")
+    }
+
+    public String getName() {
+        return "Notifications sender"
+    }
+
+    public String getStatusDescription() {
+        return status
+    }
+
+    public void trigger() {
+        // Build a multi-map of library -> list of settings objects for that library's users
+        Map<String, List<Map>> heldByToUserSettings = new HashMap<>();
+        {
+            List<Map> allUserSettingStrings = whelk.getStorage().getAllUserData()
+            for (Map settings : allUserSettingStrings) {
+                if (!settings["notificationEmail"])
+                    continue
+                settings?.requestedNotifications?.each { request ->
+                    if (!request instanceof Map)
+                        return
+                    if (!request["heldBy"])
+                        return
+
+                    String heldBy = request["heldBy"]
+                    if (!heldByToUserSettings.containsKey(heldBy))
+                        heldByToUserSettings.put(heldBy, [])
+                    heldByToUserSettings[heldBy].add(settings)
+                }
+            }
+        }
+
+        // Determine the time interval of changes for which to generate notifications.
+        Timestamp from = Timestamp.from(Instant.now().minus(1, ChronoUnit.DAYS)) // Default to last 24h if first time.
```

Comment:
Why default to 24h? Better to default to a 0 interval, i.e. first run does nothing?